server/src/controllers/NotificationController.ts

```diff
@@ -0,0 +1,108 @@
+import { Request, Response } from 'express';
+import webpush from 'web-push';
+import dotenv from 'dotenv';
+import User from '../models/User';
+
+dotenv.config();
+
+webpush.setVapidDetails(
+  'https://cs472-roadwatch.vercel.app/', // TODO: Replace with final registered domain (e.g., https://roadwatch.com )
+  process.env.NOTIFICATION_PUBLIC_KEY as string,
+  process.env.NOTIFICATION_PRIVATE_KEY as string
+);
+
+interface ISendNotificationRequestBody {
+  id: string;
+  title: string;
+  options?: NotificationOptions;
+}
+
+export const saveSubscription = async (req: Request, res: Response) => {
+  const { id, subscription } = req.body;
+
+  if (!id) {
+    return res.status(401).send({
+      error: 'Unauthorized',
+    });
+  }
+
+  try {
+    await User.findByIdAndUpdate(id, {
+      $addToSet: { notificationSubscriptions: subscription },
+    });
+
+    res.statusMessage = 'User successfully subscribed to notifications.';
+    res.sendStatus(200);
+  } catch (error) {
+    console.error('Notification Subscription Error: ', error);
+    return res.sendStatus(400);
+  }
+};
+
+// TODO: Create `sendNotificationToCommunity` endpoint.
+
+export const unsubscribe = async (
+  req: Request<object, object, { id: string; subscription: PushSubscription }>,
```

Comment:
Not sure if you're having issues with typing, but TypeScript should be smart enough to infer typing from your request. You shouldn't need to specify more than just req: Request and it would still allow you to destructure from body.

Reply:
Unfortunately it's not smart enough. I specifically did this to access some properties from the `PushSubscription` interface and to ensure it aligns with the types I assigned in `notification.slice.tsx`.

More info on what I'm doing here: https://stackoverflow.com/a/55413670
You can ignore the `object` part. That's just a way for me to declare that I don't want/need any types for any of the other request fields.